File: CodingInterview/Programmers/Python/LV1/52_K번째수.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42748?language=python3

# Number of times solving questions: 1
# 2026-08-21

import sys, pathlib
sys.path.append(str(pathlib.Path(__file__).resolve().parent.parent))  # test_runner 경로
from test_runner import run_tests

def solution(array, commands):
    # 규칙 1. 명령 하나는 [i, j, k]이고 셋 다 1부터 세는 "몇 번째"다.
    #        파이썬 인덱스는 0부터 세므로 그대로 쓰면 한 칸씩 어긋난다.
    # 규칙 2. 명령마다 자르기 -> 정렬 -> k번째 뽑기를 거치며,
    #        결과는 명령이 주어진 순서 그대로 담아 돌려준다.

    answer = []

    for i, j, k in commands:
        # 1. i번째부터 j번째까지 자른 뒤 정렬한다.
        #    시작은 i번째이므로 i-1, 끝은 j를 그대로 쓴다.
        #    슬라이스가 끝 위치를 포함하지 않기 때문에, j번째까지 포함하려면 j로 끊어야 한다. (규칙 1)
        sorted_slice = sorted(array[i-1:j])

        # 2. 정렬한 조각에서 k번째를 뽑는다. k도 1부터 세므로 k-1. (규칙 1)
        #    append 순서가 곧 반환 순서가 된다. (규칙 2)
        answer.append(sorted_slice[k-1])

    return answer

# ─────────────────────────────────────────────────────────────
# 테스트
# ─────────────────────────────────────────────────────────────
test_cases = [
    (([1, 5, 2, 6, 3, 7, 4], [[2, 5, 3], [4, 4, 1], [1, 7, 3]]), [5, 6, 3]),
]

run_tests(solution, test_cases)

File: CodingInterview/Programmers/Cpp/LV1/52_K번째수.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/42748?language=cpp
// 출처: 연습문제
// 대응 Python 풀이: ../../Python/LV1/52_K번째수.py

// Number of times solving questions: 1
// 2026-08-22

#include "../test_runner.h"
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
using namespace std;

vector<int> solution(vector<int> array, vector<vector<int>> commands) {
    vector<int> answer;

    for (const vector<int>& command : commands) {
        int i = command[0], j = command[1], k = command[2];
        vector<int> slice(array.begin() + (i-1), array.begin() + j);
        sort(slice.begin(), slice.end());
        answer.push_back(slice[k-1]);
    }

    return answer;
}

int main() {
    Tester t("solution");
    t.check(solution({1,5,2,6,3,7,4}, {{2,5,3},{4,4,1},{1,7,3}}), vector<int>{5,6,3});
    t.check("i==j", solution({1,5,2,6,3,7,4}, {{4,4,1}}), vector<int>{6});
    t.check("전체 구간 끝", solution({1,5,2,6,3,7,4}, {{1,7,7}}), vector<int>{7});
    t.check("길이 1", solution({42}, {{1,1,1}}), vector<int>{42});
    return t.report();
}
